ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpStatement.kt

```diff
@@ -0,0 +1,139 @@
+                } catch (_: Throwable) {}
+                join()
+            }
+        }
+    }
+
+    /**
+     * Executes this statement and download the response.
+     * After the method finishes, the client downloads the response body in memory and release the connection.
+     *
+     * To receive exact type you consider using [receive<T>()] method.
+     */
+    suspend fun execute(): HttpResponse = execute {
+        val savedCall = it.call.save()
+        savedCall.response
+    }
+
+    /**
+     * Executes this statement and run [HttpClient.responsePipeline] with the response and expected type [T].
+     *
+     * Note if T is a streaming type, you should manage how to close it manually.
+     */
+    @UseExperimental(ExperimentalStdlibApi::class)
+    suspend inline fun <reified T> receive(): T = when (T::class) {
+        HttpStatement::class -> this as T
+        HttpResponse::class -> execute() as T
+        else -> {
+            val response = executeUnsafe()
+            val result =  try {
+                response.receive<T>()
+            }  finally {
+                response.complete()
+            }
+
+            result
+        }
+    }
+
+    /**
+     * Executes this statement and run the [block] with a [HttpClient.responsePipeline] execution result.
+     *
+     * Note that T can be a streamed type such as [ByteReadChannel].
+     */
+    suspend inline fun <reified T, R> receive(crossinline block: suspend (response: T) -> R): R = execute {
+        val response = it.receive<T>()
+        return@execute block(response)
+    }
+
+    /**
+     * Return [HttpResponse] with open streaming body.
+     */
+    @PublishedApi
+    internal suspend fun executeUnsafe(): HttpResponse {
+        val builder = HttpRequestBuilder().takeFrom(builder)
+        @Suppress("DEPRECATION_ERROR")
+        val call = client.execute(builder)
+        return call.response
+    }
+}
+
+@Deprecated(
+    "[HttpStatement] isn't closeable.",
+    level = DeprecationLevel.ERROR,
+    replaceWith = ReplaceWith("this.execute<T>(block)")
+)
+@Suppress("KDocMissingDocumentation")
+fun <T> HttpStatement.use(block: suspend (response: HttpResponse) -> T) {
+}
+
+
+@Deprecated("", level = DeprecationLevel.ERROR, replaceWith = ReplaceWith("this.execute()"))
+@Suppress("KDocMissingDocumentation")
+val HttpStatement.response: HttpResponse
+    get() = error("")
+
+/**
+ * Read the [HttpResponse.content] as a String. You can pass an optional [charset]
+ * to specify a charset in the case no one is specified as part of the Content-Type response.
+ * If no charset specified either as parameter or as part of the response,
+ * [HttpResponseConfig.defaultCharset] will be used.
+ *
+ * Note that [charset] parameter will be ignored if the response already has a charset.
+ *      So it just acts as a fallback, honoring the server preference.
+ */
+suspend fun HttpResponse.readText(charset: Charset? = null): String {
```

Comment:
We probably can rename `charset` parameter to something like `fallbackCharset` or `defaultCharset`